### PR TITLE
Completes task 11 CORE: GET /api/articles (sorting queries)

### PR DESCRIPTION
**Pull request by Dimeben**

### CODE REVIEW

File: controllers/articles-controller.js [hunk 1]
```diff
@@ -1,9 +1,21 @@
...
 
+const validSortBy = [
+  "author",
+  "title",
+  "article_id",
+  "topic",
+  "created_at",
+  "votes",
+  "comment_count",
+];
```
Comment (nc-verity-gregory) [inline controllers/articles-controller.js:16]: While it is highly unlikely that the columns will change, a good challenge for you would be to find out how you can get the valid columns programmatically without hard coding. _Hint it only takes one google search as it is a pretty standard psql query that we haven't taught_

File: controllers/articles-controller.js [hunk 2]
```diff
@@ -16,7 +28,21 @@ exports.getArticles = (req, res, next) => {
...
 exports.getAllArticles = (req, res, next) => {
-  selectAllArticles()
+  let sortBy = "";
+  let order = "";
+  if (Object.keys(req.query).length === 0) {
+    sortBy = "created_at";
+    order = "DESC";
+  } else {
+    sortBy = req.query.sort_by;
+    order = req.query.order.toUpperCase();
```
Comment (nc-verity-gregory) [inline controllers/articles-controller.js:38]: You could use default parameters in your function to handle this. In your model you could have `selectAllArticles(sortBy = 'created_at', order='DESC')` and that would omit the need for this if statement.

File: models/articles-model.js [hunk 3]
```diff
@@ -16,21 +16,36 @@ exports.selectArticles = (articleId) => {
...
-    LEFT JOIN comments ON articles.article_id = comments.article_id
-    GROUP BY articles.article_id
-    ORDER BY articles.created_at DESC
-  `
-    )
-    .then((result) => {
-      const articles = result.rows;
-      return articles;
-    });
+exports.selectAllArticles = (sortBy, order) => {
```
Comment (nc-verity-gregory) [inline models/articles-model.js:19]: This is where you could use default parameters.

File: models/articles-model.js [hunk 4]
```diff
@@ -16,21 +16,36 @@ exports.selectArticles = (articleId) => {
...
+  GROUP BY articles.article_id`;
+
+  const validSortBy = [
+    "author",
+    "title",
+    "article_id",
+    "topic",
+    "created_at",
+    "votes",
+    "comment_count",
```
Comment (nc-verity-gregory) [inline models/articles-model.js:32]: You already have this in your controller in an attempt to whitelist, if it got through the controller is it needed here? You only need to check your request parameters in one location. As previously mentioned, while it is unlikely your data will change, it _could_ so how could you make this array more programmatic?

**Review by nc-verity-gregory — COMMENTED** [review]
It seems you aren't sure where you should be checking your parameters. I'd recommend sticking to one place as you have duplicate code in your model and controller.

The query build up is good, if you make use of some default parameters you could refactor some if conditions.

File: controllers/articles-controller.js (hunk 1 above)
Comment (Dimeben) [reply controllers/articles-controller.js:16]: Screenshot 2024-08-27 at 14 45 36


Is it that?

File: controllers/articles-controller.js (hunk 2 above)
Comment (Dimeben) [reply controllers/articles-controller.js:38]: Screenshot 2024-08-27 at 14 46 10

File: models/articles-model.js (hunk 3 above)
Comment (Dimeben) [reply models/articles-model.js:19]: Screenshot 2024-08-27 at 14 46 45

**Pull request merged**